File: api/src/main/java/io/github/mh321productions/serverapi/module/chat/standardchats/AbstractChat.java
package io.github.mh321productions.serverapi.module.chat.standardchats;

import java.util.ArrayList;
import java.util.Collection;

import io.github.mh321productions.serverapi.util.message.MessageFormatter;
import org.bukkit.entity.Player;

import io.github.mh321productions.serverapi.module.chat.Chat;
import io.github.mh321productions.serverapi.util.message.Message;

/**
 * Abstrakter Wrapper für das {@link Chat}-Interface
 * @author 321Productions
 *
 */
public abstract class AbstractChat implements Chat {
	
	protected ArrayList<Player> members;
	
	private byte actionFlags;
	private String internalName;
	
	public AbstractChat(byte flags, String name) {
		actionFlags = flags;
		internalName = name;
		members = new ArrayList<>();
	}
	
	public String getInternalName() {
		return internalName;
	}
	
	public void addMember(Player newMember) {members.add(newMember);}
	public void addMember(Collection<? extends Player> newMembers) {members.addAll(newMembers);}
	public void removeMember(Player oldMember) {members.remove(oldMember);}
	public void removeMember(Collection<? extends Player> remove) {members.removeAll(remove);}
	
	/**
	 * Formatiert die Nachricht intern (soll von {@link #sendMessage(Player, String)} aufgerufen werden)
	 * @param player Der Spieler, der die Nachricht verschickt hat
	 * @param message Die rohe Nachricht
	 * @return Die formatierte Message, die verschickt werden kann
	 */
	protected abstract Message formatMessage(Player player, String message);
	
	public boolean checkFlag(byte flag) {
		return (actionFlags & flag) != 0;
	}
	
	@Override
	public void sendMessage(Player sender, String message) {
		Message send = formatMessage(sender, message);
		MessageFormatter.sendMessage(members, send);
	}
	
	@Override
	public String toString() {
		return internalName;
	}
}

File: src/main/kotlin/io/github/mh321productions/serverapi/module/chat/standardchats/AbstractChat.kt
package io.github.mh321productions.serverapi.module.chat.standardchats

import io.github.mh321productions.serverapi.module.chat.Chat
import io.github.mh321productions.serverapi.util.message.Message
import io.github.mh321productions.serverapi.util.message.MessageFormatter.sendMessage
import org.bukkit.entity.Player
import kotlin.experimental.and

/**
 * Abstrakter Wrapper für das [Chat]-Interface
 * @author 321Productions
 */
abstract class AbstractChat(private val actionFlags: Byte, override val internalName: String) : Chat {

    protected var members = mutableListOf<Player>()

    override fun addMember(newMember: Player) {
        members.add(newMember)
    }

    override fun addMember(newMembers: Collection<Player>) {
        members.addAll(newMembers)
    }

    override fun removeMember(oldMember: Player) {
        members.remove(oldMember)
    }

    override fun removeMember(remove: Collection<Player>) {
        members.removeAll(remove)
    }

    override fun checkFlag(flag: Byte) = (actionFlags and flag) != 0.toByte()

    override fun sendMessage(player: Player, message: String) {
        val send = formatMessage(player, message)
        sendMessage(members, send)
    }

    override fun toString() = internalName

    /**
     * Formatiert die Nachricht intern (soll von [.sendMessage] aufgerufen werden)
     * @param player Der Spieler, der die Nachricht verschickt hat
     * @param message Die rohe Nachricht
     * @return Die formatierte Message, die verschickt werden kann
     */
    protected abstract fun formatMessage(player: Player, message: String): Message
}
